File: src/main/java/com/hm/weather/engine/TGALoader.java
//
// Source code recreated from a .class file by IntelliJ IDEA
// (powered by Fernflower decompiler)
//

package com.hm.weather.engine;

import com.hm.weather.engine.Utility.Logger;

import java.io.IOException;
import java.io.InputStream;
import java.nio.ByteBuffer;
import java.nio.channels.Channels;
import java.nio.channels.ReadableByteChannel;

class TGALoader {
    private static final String TAG = "GL Engine";
    private static final ByteBuffer cTGAcompare;
    private static final ByteBuffer uTGAcompare;
    private static final ByteBuffer ugTGAcompare;

    static {
        byte[] var0 = new byte[]{0, 0, 3};
        byte[] var1 = new byte[]{0, 0, 2};
        byte[] var2 = new byte[]{0, 0, 10};
        uTGAcompare = ByteBuffer.allocate(var1.length);
        uTGAcompare.put(var1);
        uTGAcompare.flip();
        cTGAcompare = ByteBuffer.allocate(var2.length);
        cTGAcompare.put(var2);
        cTGAcompare.flip();
        ugTGAcompare = ByteBuffer.allocate(var0.length);
        ugTGAcompare.put(var0);
        ugTGAcompare.flip();
    }

    TGALoader() {
    }

    private TGALoader.TGA loadCompressedTGA(ReadableByteChannel var1) throws IOException {
        Logger.v("GL Engine", " - reading compressed tga");
        TGALoader.TGA var9 = new TGALoader.TGA();
        this.readBuffer(var1, var9.header);
        var9.width = (this.unsignedByteToInt(var9.header.get(10)) << 8) + this.unsignedByteToInt(var9.header.get(9));
        var9.height = (this.unsignedByteToInt(var9.header.get(12)) << 8) + this.unsignedByteToInt(var9.header.get(11));
        var9.bpp = this.unsignedByteToInt(var9.header.get(13));
        if (var9.width > 0 && var9.height > 0 && (var9.bpp == 24 || var9.bpp == 32)) {
            var9.bytesPerPixel = var9.bpp / 8;
            var9.imageSize = var9.bytesPerPixel * var9.width * var9.height;
            var9.imageData = ByteBuffer.allocate(var9.imageSize);
            var9.imageData.position(0);
            var9.imageData.limit(var9.imageData.capacity());
            int var7 = var9.height * var9.width;
            int var3 = 0;
            int var2 = 0;
            ByteBuffer var10 = ByteBuffer.allocate(var9.bytesPerPixel);

            int var4;
            do {
                int var8;
                try {
                    ByteBuffer var11 = ByteBuffer.allocate(1);
                    var11.clear();
                    var1.read(var11);
                    var11.flip();
                    var8 = this.unsignedByteToInt(var11.get());
                } catch (IOException var12) {
                    throw new IOException("Could not read RLE header");
                }

                int var5;
                short var6;
                if (var8 < 128) {
                    var6 = 0;
                    var5 = var2;

                    while (true) {
                        var2 = var5;
                        var4 = var3;
                        if (var6 >= var8 + 1) {
                            break;
                        }

                        this.readBuffer(var1, var10);
                        var9.imageData.put(var5, var10.get(2));
                        var9.imageData.put(var5 + 1, var10.get(1));
                        var9.imageData.put(var5 + 2, var10.get(0));
                        if (var9.bytesPerPixel == 4) {
                            var9.imageData.put(var5 + 3, var10.get(3));
                        }

                        var5 += var9.bytesPerPixel;
                        ++var3;
                        if (var3 > var7) {
                            throw new IOException("Too many pixels read");
                        }

                        ++var6;
                    }
                } else {
                    this.readBuffer(var1, var10);
                    var6 = 0;
                    var5 = var2;

                    while (true) {
                        var2 = var5;
                        var4 = var3;
                        if (var6 >= var8 - 127) {
                            break;
                        }

                        var9.imageData.put(var5, var10.get(2));
                        var9.imageData.put(var5 + 1, var10.get(1));
                        var9.imageData.put(var5 + 2, var10.get(0));
                        if (var9.bytesPerPixel == 4) {
                            var9.imageData.put(var5 + 3, var10.get(3));
                        }

                        var5 += var9.bytesPerPixel;
                        ++var3;
                        if (var3 > var7) {
                            throw new IOException("Too many pixels read");
                        }

                        ++var6;
                    }
                }

                var3 = var4;
            } while (var4 < var7);

            return var9;
        } else {
            throw new IOException("Invalid header data");
        }
    }

    private TGALoader.TGA loadUncompressedTGA(ReadableByteChannel var1) throws IOException {
        Logger.v("GL Engine", " - reading uncompressed tga");
        TGALoader.TGA var4 = new TGALoader.TGA();
        this.readBuffer(var1, var4.header);
        var4.width = (this.unsignedByteToInt(var4.header.get(10)) << 8) + this.unsignedByteToInt(var4.header.get(9));
        var4.height = (this.unsignedByteToInt(var4.header.get(12)) << 8) + this.unsignedByteToInt(var4.header.get(11));
        var4.bpp = this.unsignedByteToInt(var4.header.get(13));
        if (var4.width <= 0 || var4.height <= 0 || var4.bpp != 24 && var4.bpp != 32 && var4.bpp != 8) {
            throw new IOException("Invalid header data");
        } else {
            var4.bytesPerPixel = var4.bpp / 8;
            var4.imageSize = var4.bytesPerPixel * var4.width * var4.height;
            var4.imageData = ByteBuffer.allocate(var4.imageSize);
            Logger.v("GL Engine", " - " + var4.width + "x" + var4.height + "x" + var4.bpp + "(" + var4.imageSize + ")");
            this.readBuffer(var1, var4.imageData);

            for (int var3 = 0; var3 < var4.imageSize - 2; var3 += var4.bytesPerPixel) {
                byte var2 = var4.imageData.get(var3);
                var4.imageData.put(var3, var4.imageData.get(var3 + 2));
                var4.imageData.put(var3 + 2, var2);
            }

            return var4;
        }
    }

    private void readBuffer(ReadableByteChannel var1, ByteBuffer var2) throws IOException {
        while (var2.hasRemaining()) {
            var1.read(var2);
        }

        var2.flip();
    }

    private int unsignedByteToInt(byte var1) {
        return var1 & 255;
    }

    public TGALoader.TGA loadTGA(InputStream var1) throws IOException {
        ByteBuffer var2 = ByteBuffer.allocate(3);
        ReadableByteChannel var3 = Channels.newChannel(var1);
        this.readBuffer(var3, var2);
        if (!uTGAcompare.equals(var2) && !ugTGAcompare.equals(var2)) {
            if (cTGAcompare.equals(var2)) {
                return this.loadCompressedTGA(var3);
            } else {
                var3.close();
                throw new IOException("TGA file be type 2 or type 10 ");
            }
        } else {
            return this.loadUncompressedTGA(var3);
        }
    }

    class TGA {
        int bpp;
        int bytesPerPixel;
        int commentLength;
        ByteBuffer header = ByteBuffer.allocate(15);
        int height;
        ByteBuffer imageData;
        int imageSize;
        final TGALoader loader = TGALoader.this;
        int type;
        int width;

        public TGA() {
        }
    }
}

File: app/src/main/kotlin/io/github/gmazzo/android/livewallpaper/weather/engine/models/TGALoader.kt
//
// Source code recreated from a .class file by IntelliJ IDEA
// (powered by Fernflower decompiler)
//
package io.github.gmazzo.android.livewallpaper.weather.engine.models

import android.util.Log
import java.io.IOException
import java.io.InputStream
import java.nio.ByteBuffer
import java.nio.channels.Channels
import java.nio.channels.ReadableByteChannel

internal class TGALoader {
    @Throws(IOException::class)
    private fun loadCompressedTGA(var1: ReadableByteChannel): TGA {
        Log.v("GL Engine", " - reading compressed tga")
        val var9 = TGA()
        this.readBuffer(var1, var9.header)
        var9.width = (this.unsignedByteToInt(var9.header[10]) shl 8) + this.unsignedByteToInt(
            var9.header[9]
        )
        var9.height = (this.unsignedByteToInt(var9.header[12]) shl 8) + this.unsignedByteToInt(
            var9.header[11]
        )
        var9.bpp = this.unsignedByteToInt(var9.header[13])
        if (var9.width > 0 && var9.height > 0 && (var9.bpp == 24 || var9.bpp == 32)) {
            var9.bytesPerPixel = var9.bpp / 8
            var9.imageSize = var9.bytesPerPixel * var9.width * var9.height
            var9.imageData = ByteBuffer.allocate(var9.imageSize).apply {
                position(0)
                limit(capacity())
            }
            val var7 = var9.height * var9.width
            var var3 = 0
            var var2 = 0
            val var10 = ByteBuffer.allocate(var9.bytesPerPixel)

            var var4: Int
            do {
                var var8: Int
                try {
                    val var11 = ByteBuffer.allocate(1)
                    var11.clear()
                    var1.read(var11)
                    var11.flip()
                    var8 = this.unsignedByteToInt(var11.get())
                } catch (var12: IOException) {
                    throw IOException("Could not read RLE header")
                }

                var var5: Int
                var var6: Short
                if (var8 < 128) {
                    var6 = 0
                    var5 = var2

                    while (true) {
                        var2 = var5
                        var4 = var3
                        if (var6 >= var8 + 1) {
                            break
                        }

                        this.readBuffer(var1, var10)
                        var9.imageData!!
                            .put(var5, var10[2])
                            .put(var5 + 1, var10[1])
                            .put(var5 + 2, var10[0])
                        if (var9.bytesPerPixel == 4) {
                            var9.imageData!!.put(var5 + 3, var10[3])
                        }

                        var5 += var9.bytesPerPixel
                        ++var3
                        if (var3 > var7) {
                            throw IOException("Too many pixels read")
                        }

                        ++var6
                    }
                } else {
                    this.readBuffer(var1, var10)
                    var6 = 0
                    var5 = var2

                    while (true) {
                        var2 = var5
                        var4 = var3
                        if (var6 >= var8 - 127) {
                            break
                        }

                        var9.imageData!!
                            .put(var5, var10[2])
                            .put(var5 + 1, var10[1])
                            .put(var5 + 2, var10[0])
                        if (var9.bytesPerPixel == 4) {
                            var9.imageData!!.put(var5 + 3, var10[3])
                        }

                        var5 += var9.bytesPerPixel
                        ++var3
                        if (var3 > var7) {
                            throw IOException("Too many pixels read")
                        }

                        ++var6
                    }
                }

                var3 = var4
            } while (var4 < var7)

            return var9
        } else {
            throw IOException("Invalid header data")
        }
    }

    @Throws(IOException::class)
    private fun loadUncompressedTGA(var1: ReadableByteChannel): TGA {
        Log.v("GL Engine", " - reading uncompressed tga")
        val var4 = TGA()
        this.readBuffer(var1, var4.header)
        var4.width = (this.unsignedByteToInt(var4.header[10]) shl 8) + this.unsignedByteToInt(
            var4.header[9]
        )
        var4.height = (this.unsignedByteToInt(var4.header[12]) shl 8) + this.unsignedByteToInt(
            var4.header[11]
        )
        var4.bpp = this.unsignedByteToInt(var4.header[13])
        if (var4.width <= 0 || var4.height <= 0 || var4.bpp != 24 && var4.bpp != 32 && var4.bpp != 8) {
            throw IOException("Invalid header data")
        } else {
            var4.bytesPerPixel = var4.bpp / 8
            var4.imageSize = var4.bytesPerPixel * var4.width * var4.height
            var4.imageData = ByteBuffer.allocate(var4.imageSize)
            Log.v(
                "GL Engine",
                " - " + var4.width + "r" + var4.height + "r" + var4.bpp + "(" + var4.imageSize + ")"
            )
            this.readBuffer(var1, var4.imageData)

            var var3 = 0
            while (var3 < var4.imageSize - 2) {
                val var2 = var4.imageData!!.get(var3)
                var4.imageData!!
                    .put(var3, var4.imageData!!.get(var3 + 2))
                    .put(var3 + 2, var2)
                var3 += var4.bytesPerPixel
            }

            return var4
        }
    }

    @Throws(IOException::class)
    private fun readBuffer(var1: ReadableByteChannel, var2: ByteBuffer?) {
        while (var2!!.hasRemaining()) {
            var1.read(var2)
        }

        var2.flip()
    }

    private fun unsignedByteToInt(var1: Byte): Int {
        return var1.toInt() and 255
    }

    @Throws(IOException::class)
    fun loadTGA(var1: InputStream?): TGA {
        val var2 = ByteBuffer.allocate(3)
        val var3 = Channels.newChannel(var1)
        this.readBuffer(var3, var2)
        return if (uTGAcompare != var2 && ugTGAcompare != var2) {
            if (cTGAcompare == var2) {
                loadCompressedTGA(var3)
            } else {
                var3.close()
                throw IOException("TGA file be type 2 or type 10 ")
            }
        } else {
            loadUncompressedTGA(var3)
        }
    }

    internal inner class TGA {
        var bpp: Int = 0
        var bytesPerPixel: Int = 0
        var commentLength: Int = 0
        var header: ByteBuffer = ByteBuffer.allocate(15)
        var height: Int = 0
        var imageData: ByteBuffer? = null
        var imageSize: Int = 0
        val loader: TGALoader = this@TGALoader
        var type: Int = 0
        var width: Int = 0
    }

    companion object {
        private const val TAG = "GL Engine"
        private val cTGAcompare: ByteBuffer
        private val uTGAcompare: ByteBuffer
        private val ugTGAcompare: ByteBuffer

        init {
            val var0 = byteArrayOf(0, 0, 3)
            val var1 = byteArrayOf(0, 0, 2)
            val var2 = byteArrayOf(0, 0, 10)
            uTGAcompare = ByteBuffer.allocate(var1.size)
            uTGAcompare.put(var1)
            uTGAcompare.flip()
            cTGAcompare = ByteBuffer.allocate(var2.size)
            cTGAcompare.put(var2)
            cTGAcompare.flip()
            ugTGAcompare = ByteBuffer.allocate(var0.size)
            ugTGAcompare.put(var0)
            ugTGAcompare.flip()
        }
    }
}
